### DISCUSSION TITLE
Remapping alt + tab to ctrl + tab

### DISCUSSION THREAD

**Discussion by yaserahmady** [cajhin/capsicain#68] [Q&A]
Hey there, thank you so much for capsicain: it makes me miss macOS a little more. [This is my whole config](https://github.com/yaserahmady/miss-you) if anyone reading from the future finds it useful.

I was wondering if there's a way to remap <kbd>alt</kbd> + <kbd>tab</kbd> with capsicain only. My goal is to mimic macOS' <kbd>⌘ cmd</kbd> + <kbd>tab</kbd> shortcut.

| From                                               | To                                                  |
|----------------------------------------------------|-----------------------------------------------------|
| <kbd>alt</kbd> + <kbd>tab</kbd>                    | <kbd>ctrl</kbd> + <kbd>tab</kbd>                    |
| <kbd>alt</kbd> + <kbd>shift</kbd> + <kbd>tab</kbd> | <kbd>ctrl</kbd> + <kbd>shift</kbd> + <kbd>tab</kbd> |

I was able to kinda do it in AutoHotKey. I tried a couple of ways but they all end up randomly not releasing the keys anymore which screws up the whole typing experience until I'm able to <kbd>esc</kbd> + <kbd>y</kbd> or close AutoHotKey from the tray icon.

[The four methods I tried](https://github.com/yaserahmady/miss-you/tree/main/alt-tab) (each one sucks in its own way).

I have a Logitech MX Keys with an Italian keyboard layout:

![mx-t](https://user-images.githubusercontent.com/51824028/150974141-f1e3646c-9908-4962-b560-7befd71df9ad.png)

My fairly simple `capsicain.ini`:

```ini
GLOBAL StartAHK
GLOBAL StartInTraybar
GLOBAL ActiveConfigOnStartup 1

[CONFIG_1]
OPTION configName macOS
REWIRE LALT LCTRL
REWIRE LCTRL LWIN
REWIRE LWIN LALT

# COMBO TAB [..&&] > key( TAB + &..& )
# COMBO TAB [..&.] > key( TAB + &... )
# COMBO ` [..&.] > key('')

[CONFIG_2]
OPTION configName Gaming
```

**Comment by cajhin** [comment]
Sorry, bad news: 
1. you cannot properly map Alt+Tab to Ctrl+Tab with capsicain (and I think AHK won't do it either)
2. I don't really understand what you're trying to achieve here. Flipping Alt and Ctrl does not make an Apple keyboard?
CMD+Tab in MacOS is equivalent in Windows to Alt+Tab, not to Ctrl+Tab (switch apps. Ctrl+tab switches tabs).

fyi this kinda does what you requested:
combo tab [&...] > moddedKey( tab + &. )   #alt+tab to ctrl+tab

moddedKey() solves the issue that, when you trigger the combo ALT+TAB, then at this point Windows already knows ALT is down.
capsicain releases ALT, sends CTRL+TAB, then presses ALT again.

Now your mapping works BUT 
for the special case with TAB, you don't want Ctrl+Tab, Ctrl+Tab, Ctrl+Tab,
you want Ctrl+Tab Tab Tab....
Capsicain (and AHK?) has no rule that lets you say "when Alt+Tab is pressed, make it Ctrl+Tab, and then go on pretending that Alt is Ctrl until I release it"

**Reply by yaserahmady** [reply]
Thanks for replying! What I actually wanted was to switch the behaviour of <kbd>alt</kbd> + <kbd>tab</kbd> and <kbd>ctrl</kbd> + <kbd>tab</kbd>, so that:

- <kbd>ctrl</kbd> + <kbd>tab</kbd> switches apps
- <kbd>alt</kbd> + <kbd>tab</kbd> switches whatever (it's not really important for me since I didn't have this on macOS)

> I don't really understand what you're trying to achieve here. Flipping Alt and Ctrl does not make an Apple keyboard?
CMD+Tab in MacOS is equivalent in Windows to Alt+Tab, not to Ctrl+Tab (switch apps. Ctrl+tab switches tabs).

Coming from macOS I thought that if in Windows I `REWIRE LALT LCTRL` then I'll basically get most macOS shortcuts on Windows for free without the need of remapping every single keyboard shortcut in AutoHotKey.

The biggest problem is that on macOS for windows switching you use <kbd>⌘ cmd</kbd> + <kbd>tab</kbd> and not <kbd>alt</kbd> + <kbd>tab</kbd>. Now in Windows with my remaps, since I `REWIRE LALT LCTRL` I swapped <kbd>alt</kbd> with <kbd>ctrl</kbd> so to get the same key position as on an Apple keyboard I should find a way to make <kbd>ctrl</kbd> + <kbd>tab</kbd> behave like <kbd>alt</kbd> + <kbd>tab</kbd>.

My strategy was to make Windows' <kbd>ctrl</kbd> act like <kbd>⌘ cmd</kbd> on macOS, it's proving really hard to do!

> you cannot properly map Alt+Tab to Ctrl+Tab with capsicain (and I think AHK won't do it either)

That's unfortunate… Thank you for your help anyways!

---

I'm gonna document the ways I tried to solve it so if anyone finds this from a search: this is what I tried with AutoHotKey.

### [Method 1](https://github.com/yaserahmady/miss-you/blob/main/alt-tab/method-01.ahk)

This kinda works but <kbd>shift</kbd> + <kbd>ctrl</kbd> + <kbd>tab</kbd> is impossible to do simply because the special event `ShiftAltTab` doesn't support more than 2 keys.

```ahk
; Docs: https://www.autohotkey.com/docs/Hotkeys.htm#AltTabDetail
; Each Alt-Tab hotkey must be either a single key or a combination of two keys, which is typically achieved via the ampersand symbol (&).
; Currently, all special Alt-tab actions must be assigned directly to a hotkey as in the examples above (i.e. they cannot be used as though they were commands). They are not affected by #IfWin or #If.
; An alt-tab action may take effect on key-down and/or key-up regardless of whether the up keyword is used, and cannot be combined with another action on the same key. For example, using both F1::AltTabMenu and F1 up::OtherAction() is unsupported.

<^Tab::AltTab
<+Tab::ShiftAltTab
```

### [Method 2](https://github.com/yaserahmady/miss-you/blob/main/alt-tab/method-02.ahk)

This randomly misses some key up events so one of your keys stays pressed which messes up every other keyboard shortcut you press until you're able to quit AutoHotKey.

```ahk
; https://superuser.com/a/1248610

LCtrl & Tab:: 
    AltTabMenu := true
    If GetKeyState("Shift", "P")
        Send, {Alt Down}{Shift Down}{Tab}
    else
        Send, {Alt Down}{Tab}
return

if (AltTabMenu) {
~*LCtrl Up::
    Send, {Shift Up}{Alt Up}
    AltTabMenu := false 
return
}
```

### [Method 3](https://github.com/yaserahmady/miss-you/blob/main/alt-tab/method-03.ahk)
Same issues as Method 2.

```ahk
; https://www.autohotkey.com/boards/viewtopic.php?p=308245#p308245

; Remap Ctrl-Tab to Alt-Tab
^Tab::
    Send {Alt down}{Tab}
    Keywait Control
    Send {Alt up}
return

; Remap Ctrl-Shift-Tab to Alt-Shift-Tab
^+Tab::
    Send {Alt down}{Shift down}{Tab}
    Keywait Control
    Send {Alt up}
    Send {Shift up}
return
```

### [Method 4](https://github.com/yaserahmady/miss-you/blob/main/alt-tab/method-04.ahk)

This one actually works ok but it's a full alt-tab remake with AutoHotKey. It would be perfect but modifying it to work with <kbd>ctrl</kbd> instead of <kbd>alt</kbd> is kinda hard since the code requires some advanced knowledge of AutoHotKey and Windows API (which I don't have). Still I got it to work kinda ok but I'd rather just switch <kbd>ctrl</kbd> + <kbd>tab</kbd> with <kbd>alt</kbd> + <kbd>tab</kbd> and have the native UI.
